### ISSUE TITLE
spec.satisfies needs to be refactored (behaves wrongly inside packages)

### ISSUE THREAD

**Issue by alalazo** [spack/spack#4135]
This may be the root of some weird behaviors seen around, given that currently this pattern is used in more than 40 packages. Short story: ~never use this inside packages:~  this line behaves wrongly in packages:
```python
spec.satisfies('^foo')  # or similar checks
```
while this behaves correctly:
```python
'^foo' in spec
```

The point is that the first call is meant to be used inside the concretizer and has a weaker semantic on enforcing the constraint. The second call has a stronger semantic, and is equivalent to calling:
```
for s in self.traverse():
    if s.satisfies('^foo', strict=True):
        return True
```
where:
1. we recurse on the spec
2. we activate the `strict=True` parameter which enforces the strong semantic

In other words (@tgamblin :tm: ) the first call means that `spec` **could satisfy** `^foo` eventually. The second that `spec` can satisfy `^foo` right now.

@scheibelp @becker33 @tgamblin Can you confirm that this could be a potential issue for packages?

**Comment by alalazo** [comment]
For the record, I encountered this today trying to build:
```console
$ spack spec -I cp2k@4.1+mpi ^intel-mkl ^intel-mpi % intel
Input spec
--------------------------------
     cp2k@4.1+mpi
         ^intel-mkl
         ^intel-mpi%intel

Normalized
--------------------------------
     cp2k@4.1+mpi
         ^elpa
             ^intel-mkl
             ^intel-mpi%intel
         ^fftw
         ^libint@:1.2
             ^autoconf@2.52:
                 ^m4@1.4.6:
             ^automake
             ^libtool
         ^pexsi+fortran
             ^parmetis
                 ^cmake@2.8:
                 ^metis@5:
         ^python
             ^bzip2
             ^ncurses
                 ^pkg-config
             ^openssl
                 ^zlib
             ^readline
             ^sqlite

Concretized
--------------------------------
     cp2k@4.1%intel@17.0.2+mpi~plumed arch=linux-rhel6-x86_E5v1_IntelIB 
         ^elpa@2016.05.004%intel@17.0.2~openmp arch=linux-rhel6-x86_E5v1_IntelIB 
             ^intel-mkl@2017.2.174%intel@17.0.2~ilp64~openmp+shared arch=linux-rhel6-x86_E5v1_IntelIB 
             ^intel-mpi@2017.2.174%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
         ^fftw@3.3.6-pl2%intel@17.0.2+double+float+long_double+mpi~openmp~pfft_patches~quad arch=linux-rhel6-x86_E5v1_IntelIB 
         ^libint@1.1.6%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
             ^autoconf@2.69%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
                 ^m4@1.4.18%intel@17.0.2+sigsegv arch=linux-rhel6-x86_E5v1_IntelIB 
                     ^libsigsegv@2.11%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
             ^automake@1.15%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
             ^libtool@2.4.6%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
         ^pexsi@0.9.2%intel@17.0.2+fortran arch=linux-rhel6-x86_E5v1_IntelIB 
             ^parmetis@4.0.3%intel@17.0.2~debug~gdb+shared arch=linux-rhel6-x86_E5v1_IntelIB 
                 ^cmake@3.8.0%intel@17.0.2~doc+ncurses+openssl+ownlibs~qt arch=linux-rhel6-x86_E5v1_IntelIB 
                     ^ncurses@6.0%intel@17.0.2~symlinks arch=linux-rhel6-x86_E5v1_IntelIB 
                         ^pkg-config@0.23%intel@17.0.2+internal_glib arch=linux-rhel6-x86_E5v1_IntelIB 
                     ^openssl@system%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
                 ^metis@5.1.0%intel@17.0.2~debug~gdb~int64~real64+shared arch=linux-rhel6-x86_E5v1_IntelIB 
             ^superlu-dist@4.3%intel@17.0.2~int64 arch=linux-rhel6-x86_E5v1_IntelIB 
         ^python@2.7.13%intel@17.0.2~tk~ucs4 arch=linux-rhel6-x86_E5v1_IntelIB 
             ^bzip2@1.0.6%intel@17.0.2+shared arch=linux-rhel6-x86_E5v1_IntelIB 
             ^readline@7.0%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
             ^sqlite@3.18.0%intel@17.0.2 arch=linux-rhel6-x86_E5v1_IntelIB 
             ^zlib@1.2.11%intel@17.0.2+pic+shared arch=linux-rhel6-x86_E5v1_IntelIB
```
failing on `elpa` because it was trying to link at configure time with `libmkl_blacs_openmpi`. Which is due to the check in `intel-mkl`:
```python
if self.spec.satisfies('^openmpi'):
    libnames.append('libmkl_blacs_openmpi')
```
being resolved as true.

**Comment by davydden** [comment]
>  Short story: never use this inside packages:

wow, we should add this to documentation and big bold letters! who would have thought...

**Comment by citibeth** [comment]
OK, this is really subtle... It looks like a booby trap to me. I suggest we think through ways to get rid of the booby trap, instead of just documenting it.  For example:

1. Change the default for `spec.satisfies` from `strict=False` to `strict=True`.  Then users can blithely use it in `package.py` without repercussions.

2. Rename `spec.satisfies` to `spec._satisfies`, it is therefore marked for internal use only.

3. Remove the `strict` parameter from `spec.satisfies()`, and have the behavior of the new `spec.satisfies` default to `strict=True`.  Then create a new `spec.satisfies_permissive()`, which is equivalent to `strict=False`.

**Comment by tgamblin** [comment]
I agree with @citibeth that this is a bug that should be fixed rather than advice we should give to packagers.  I'll look at what it would take to change.

**Comment by tgamblin** [comment]
The main issue here is that satisfies() was written to be used in concretization, where you want to know whether something *could* be satisfied, and it was repurposed for use outside that method, where you want to know whether something *is* satisfied.  In retrospect, there should really be two methods.

**Comment by alalazo** [comment]
I agree with @citibeth except on point 3. I think we should not augment the redundancy in the public API. If:
```python
'^foo' in spec
```
is equivalent to
```python
spec.satisfies('^foo')
```
then one of the two must disappear in the long run. 

If instead we  want a function that peeks only at *this node* in the DAG we should rename `satisfy` to something else. I don't think anybody used it on purpose to search only the current node in the DAG without traversing dependencies.

**Comment by davydden** [comment]
> If instead we want a function that peeks only at this node in the DAG

is not this the most common usage case, see which variants are enabled for a single node in DAG which is being installed? I.e. '+debug', or '+mpi'?

**Comment by alalazo** [comment]
@davydden I think the fact is if you use:
```python
'+mpi' in spec
```
in a package you expect to find it at this node. But what you are really saying is:
```python
for s in self.traverse():
    if s.satisfies('+mpi', strict=True):
        return True
```
I'll try to play with unit test to see if I find interesting corner cases. One I am thinking right now is:
```python
spec = Spec(foo~openmp ^libfoo+openmp)
assert '+openmp' in spec # ?
assert '~openmp' in spec # ?
```

Instead when you say:
```python
'^foo' in spec
```
you don't want to know if it is a direct dependency or not (at least I don't want it). You just want to know if you depend on `foo` in any way.

**Comment by citibeth** [comment]
>
> I agree with @citibeth <https://github.com/citibeth> except on point 3.
>

I was throwing out 1,2, and 3 as brainstorm alternatives.  I would not
expect we implement more than one of them.


> I think we should not augment the redundancy in the public API. If:
>
> '^foo' in spec
>
> is equivalent to
>
> spec.satisfies('^foo')
>
> then one of the two must disappear in the long run.
>

I think that `spec.satisfies()` feels more clear and explicit.  But using
`__contains__()` is so convenient...  If I had to vote for getting rid of
one, I'd get rid of `__contains__()` because it's less clear.

The main issue here is that satisfies() was written to be used in
> concretization, where you want to know whether something *could* be
> satisfied,


Then maybe it's mis-named, and the version used in concretization should be
called `could_satisfy()`.

**Comment by davydden** [comment]
> in a package you expect to find it at this node. But what you are really saying is:

then it means that
```
if '+debug' in spec
```
will return `true` if there is any dependency package compiled with `+debug` variant, whereas what one is really interested in is if **this specific** package has `+debug` variant enabled. 

I would say 99% `if '+variant' in spec` is used to check this specific package.

**Comment by alalazo** [comment]
> I would say 99% if '+variant' in spec is used to check this specific package.

I suspect this works right now because packagers / users are sane people and don't do things like:
```
foo+debug%gcc ^bar~debug%intel
```
but as I said, the code is complex enough for me that I need to check it in a debugger to be sure of what I am saying here.

**Comment by tgamblin** [comment]
I would suggest these semantics:

1. `spec.satisfies(other)`: same as current satisfies, but always strict. Concretizer uses `could_satisfy`.  I actually have a branch where I started that refactor.
2. `other in spec`: I like this notation a lot, and I think it makes the code more readable, both when you're checking whether some package is in the DAG, and when you're checking whether the spec itself satisfies an anonymous constraint.  My intuitive semantics for this would be: 
    ```python
    if other.name:
        return any(d.satisfies(other) for d in spec.traverse()
    else:
        return spec.satisfies(other)
    ```
3. `could_satisfy`: new method for use by the concretizer, equivalent to `satisfies` with `strict=False`.

This makes `satisfies` do what you'd expect, and it makes `__contains__` clear because it *only* applies to dependencies when you're explicit about which one.  If you want to check an anonymous spec against everything in the DAG, you have to be explicit and say:

```python
    any('+debug' in d for d in spec)    # or
    any(d.satisfies('+debug') for d in spec)
```

**Comment by tgamblin** [comment]
Another question kind of brought up by the last one: should `Spec` have an `__iter__` method equivalent to `spec.traverse()` with no arguments?  It would allow:

```python
    any('+debug' in d for d in spec)
```

You would still need `spec.traverse()` for when you want something other than the default traversal over dependencies (different order, etc.)

**Comment by alalazo** [comment]
@tgamblin I'll try to submit `xfail` tests to code against, and change the title of the issue to reflect the decisions here. 

One thing I don't understand:
```python
if other.name:
    return any(d.satisfies(other) for d in spec.traverse()
else:
    return spec.satisfies(other)
```
here `other` is not `autospec`ed (otherwise we will always fall in the `True` part of the branch). 

Then if I am coding package `Bar`:
```python
# True if foo is a direct dependency of bar
if '^foo' in spec:
...
# True if foo is in the DAG anywhere
if 'bar ^foo' in spec:
...
```

In case I am not missing the behavior above,  I find it quite counter-intuitive given that I took care of putting `bar` next to `^foo`. I would expect they mean the same thing at least.

**Comment by citibeth** [comment]
On Fri, May 5, 2017 at 12:43 PM, Todd Gamblin <notifications@github.com>
wrote:

> Another question kind of brought up by the last one: should Spec have an
> __iter__ method equivalent to spec.traverse() with no arguments? It would
> allow:
>
>     any('+debug' in d for d in spec)
>
> Yes

My only question is, does this add some redundancy. Wouldn't:

*any(d.name <http://d.name>=='x' for d in spec)*


be the same as:

*'^x' in d*


-- Elizabeth

**Comment by tgamblin** [comment]
@citibeth: I think something got lost in translation there.  Can you edit the comment?

**Comment by melven** [comment]
I have a similar problem in environments.

Without activate environment, I can build a spec including ^intel-mkl ^openmpi that requires scalapack/blacs libraries.

When just using "spack install" with an active environment, scalapack_libs in lib/spack/spack/build_systems/intel.py raises 'Cannot find a BLACS library for the given MPI.'.

(current develop branch)

**Comment by scheibelp** [comment]
@melven this issue predates the environments feature: can you open a new issue including the commands, environment, and stack trace (when running with `spack -d`)?

**Comment by alalazo** [comment]
Finally done in #35681
